### PR TITLE
hotfix for memory leak in aot autograd induced by saving tensors for backward

### PR DESCRIPTION
**Pull request by bdhirsh**
Workaround fix in AOTAutograd for https://github.com/pytorch/pytorch/issues/94990 (see the comments for more details / discussion)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95392
* #95421
* __->__ #95101

### CODE REVIEW

**Review by Chillee — COMMENTED** [review]
I'm worried this will have performance ramifications - we're running a detach call per activation saved.

**Comment by albanD** [comment]
> I'm worried this will have performance ramifications - we're running a detach call per activation saved.

We do a lot of detach internally haha, I really don't think this will change anything unless some FakeTensorMode weirdness kicks in,.

**Review by albanD — APPROVED** [review]
SGTM

**Comment by bdhirsh** [comment]
I'll try to get some benchmarks just to be safe

**Comment by bdhirsh** [comment]
Some sanity benchmarks:

`x.detach()` cost: ~7us (per tensor)

cost of running just the forward of a no-op AOTAutograd function: ~110us

cost of running the forward + backward of a no-op AOTAutograd function: ~540us

no-op code:
```
from functorch.compile import aot_function, nop

def f(x):
    return x

a = torch.ones(1, device='cuda', requires_grad=True)
f_ = aot_function(f, fw_compiler=nop)

t0 = benchmark.Timer(
    stmt='out = f_(a); out.sum().backward()',
    globals={'f_': f_, 'a': a}
)
print(t0.timeit(100))
```

Alban also convinced me that trying to fix this ref cycle directly in autograd (aka option b that I put in the issue) is pretty non-trivial / might require doing some extra calculations in the general case eager hot-path :(. LMK if anyone has strong opinions. That leads me towards taking on the minor overhead for now for the release, and ideally fixing this ref cycle using option (a) in the longer term.

I also need to make some other fixes since this PR broke tests - I need to properly fix up the requires_grad field of the saved tensors in the backward, so we don't lose the nice error message that you're supposed to get when you try to double backward.

**Comment by Chillee** [comment]
> cost of running just the forward of a no-op AOTAutograd function: ~110us

Jeez. That seems pretty bad lol. I think this has gotten worse 🤔

File: test/functorch/test_aotdispatch.py [hunk 1]
```diff
@@ -1016,6 +1016,34 @@ def inp_callable(req_grad):
...
+    @unittest.skipIf(not torch.cuda.is_available(), "CUDA is unavailable")
+    def test_mem_leak_from_save_for_bw(self):
+        # See a full diagnosis at this issue: https://github.com/pytorch/pytorch/issues/94990
+        # Note [Detaching saved tensors in AOTAutograd]
+        # This program creates a ref-cycle. Long term, we should fix this ref cycle
+        # (since it can arise, naturally albeit rarely, from uses of autograd.Function).
+        # But AOTAutograd makes it more likely to show up from tracing user programs,
+        # so we deal with it by manually detaching the tensors that we save for backward.
+        # This is completely wrong and would give wrong results if we were to do double backward.
+        # Fortunately today, double backward is explicitly banned in AOTAutograd.
```
Comment (gchanan) [inline test/functorch/test_aotdispatch.py:1028]: would we obviously trigger this issue if we tried to run double backwards?

Comment (bdhirsh) [reply test/functorch/test_aotdispatch.py:1028]: just checking- you're asking if whether or not we will fail loudly with a nice error message if you try to run double backwards today (and not e.g. silently get wrong results / get a bad error message)?

Jeffrey did some work a few weeks ago to make sure we get a nice error message. I actually just updated this PR to make sure that we still get the nice error (there's an expect test that asserts we get a good error message today).

Comment (bdhirsh) [reply test/functorch/test_aotdispatch.py:1028]: Oh, you're wondering if this ref cycle can show up more commonly if you're doing double backward with an `autograd.Function`?

I'm not 100% sure, but I'm leaning towards no - only this specific usage of autograd.Function will cause the mem leak (regardless of whether you're doing single or double backward).

File: torch/_functorch/aot_autograd.py [hunk 2]
```diff
@@ -2180,7 +2180,8 @@ def forward(ctx, *deduped_flat_tensor_args):
                 assert all(
                     [isinstance(x, torch.Tensor) for x in tensors_saved_for_backwards]
                 )
-                ctx.save_for_backward(*tensors_saved_for_backwards)
+                # See Note [Detaching saved tensors in AOTAutograd]
+                ctx.save_for_backward(*map(lambda x: x.detach() if x._is_view() else x, tensors_saved_for_backwards))
```
Comment (bdhirsh) [inline torch/_functorch/aot_autograd.py:2184]: @Chillee this should now be a bit less egregious, and still fixes the mem leak 🙂 (we only need to worry about the case when our saved-for-bw tensor is a differentiable view)
